src/interface.jl

```diff
@@ -440,57 +471,152 @@ logabsdetjac(b::Logit{<:Real}, x) = @. - log((x - b.a) * (b.b - x) / (b.b - b.a)
 
 #################
 # Shift & Scale #
 #################
-struct Shift{T} <: Bijector
+struct Shift{T, N} <: Bijector{N}
     a::T
 end
 
+Shift(a::T; dim::Type{Val{D}} = Val{0}) where {T<:Real, D} = Shift{T, D}(a)
+Shift(a::A; dim::Type{Val{D}} = Val{N}) where {T, D, N, A<:AbstractArray{T, N}} = Shift{A, N}(a)
+
 (b::Shift)(x) = b.a + x
 (b::Shift{<:Real})(x::AbstractArray) = b.a .+ x
 (b::Shift{<:AbstractVector})(x::AbstractMatrix) = b.a .+ x
 
-inv(b::Shift) = Shift(-b.a)
-logabsdetjac(b::Shift, x) = zero(eltype(x))
-# FIXME: ambiguous whether or not this is actually a batch or whatever
-logabsdetjac(b::Shift{<:Real}, x::AbstractMatrix) = zeros(eltype(x), size(x, 2))
-logabsdetjac(b::Shift{<:AbstractVector}, x::AbstractMatrix) = zeros(eltype(x), size(x, 2))
+inv(b::Shift{T, N}) where {T, N} = Shift{T, N}(-b.a)
+
+# FIXME: implement custom adjoint to ensure we don't get tracking
+logabsdetjac(b::Shift{T, N}, x) where {T, N} = _logabsdetjac_shift(b.a, x, Val{N})
+
+_logabsdetjac_shift(a::Real, x::Real, ::Type{Val{0}}) = zero(eltype(x))
+_logabsdetjac_shift(a::Real, x::AbstractVector{T}, ::Type{Val{0}}) where {T<:Real} = zeros(T, length(x))
+_logabsdetjac_shift(a::T1, x::AbstractVector{T2}, ::Type{Val{1}}) where {T1<:Union{Real, AbstractVector}, T2<:Real} = zero(T2)
+_logabsdetjac_shift(a::T1, x::AbstractMatrix{T2}, ::Type{Val{1}}) where {T1<:Union{Real, AbstractVector}, T2<:Real} = zeros(T2, size(x, 2))
+
+function _logabsdetjac_shift(a::TrackedReal, x::Real, ::Type{Val{0}})
+    return Tracker.param(_logabsdetjac_shift(data(a), data(x), Val{0}))
+end
+function _logabsdetjac_shift(a::TrackedReal, x::AbstractVector{T}, ::Type{Val{0}}) where {T<:Real}
+    return Tracker.param(_logabsdetjac_shift(data(a), data(x), Val{0}))
+end
+function _logabsdetjac_shift(a::T1, x::AbstractVector{T2}, ::Type{Val{1}}) where {T1<:Union{TrackedReal, TrackedVector}, T2<:Real}
+    return Tracker.param(_logabsdetjac_shift(data(a), data(x), Val{1}))
+end
+function _logabsdetjac_shift(a::T1, x::AbstractMatrix{T2}, ::Type{Val{1}}) where {T1<:Union{TrackedReal, TrackedVector}, T2<:Real}
+    return Tracker.param(_logabsdetjac_shift(data(a), data(x), Val{1}))
+end
+
 
-struct Scale{T} <: Bijector
+struct Scale{T, N} <: Bijector{N}
     a::T
 end
 
-(b::Scale)(x) = b.a * x
+Scale(a::T; dim::Type{Val{D}} = Val{0}) where {T<:Real, D} = Scale{T, D}(a)
+Scale(a::A; dim::Type{Val{D}} = Val{N}) where {T, D, N, A<:AbstractArray{T, N}} = Scale{A, D}(a)
+
+(b::Scale)(x) = b.a .* x
 (b::Scale{<:Real})(x::AbstractArray) = b.a .* x
-(b::Scale{<:AbstractVector{<:Real}})(x::AbstractMatrix{<:Real}) = x * b.a
+(b::Scale{<:AbstractMatrix})(x::AbstractArray) = b.a * x
+(b::Scale{<:AbstractVector{<:Real}, 2})(x::AbstractMatrix{<:Real}) = b.a .* x
+
+inv(b::Scale{T, D}) where {T, D} = Scale(inv(b.a); dim = Val{D})
+inv(b::Scale{<:AbstractVector, D}) where {D} = Scale(inv.(b.a); dim = Val{D})
+
+# We're going to implement custom adjoint for this
```

Comment:
I didn't check the adjoint implementations. I'm fine with then as long as we have test of them aginst AD based.

Reply:
Good point; I'll add that :+1:

Also the reason why I've added these is because of type-issues arising when one input is non-tracked and parameters potentially are tracked, etc. Difficult to know which type to `zero(T)`.

Reply:
Done